### PR TITLE
feat(kuma-cp): improve BuildRules algorithm

### PR DESCRIPTION
**Pull request by lobkovilya**
While implementing `targetRef.Kind: MeshHTTPRoute` I've encountered the problem when `BuildRules` returns excessive results because I'm adding an artificial tag `__route_hash__` to the Subset. 

But the problem could be noticed even for the given MeshTrafficPermission policy:

```yaml
type: MeshTrafficPermission
spec:
  targetRef:
    kind: Mesh
  from:
    - targetRef:
        kind: MeshService
        name: frontend
      default:
        action: Allow
    - targetRef:
        kind: MeshServiceSubset
        name: backend
        tags:
          version: v1
      default:
        action: Deny
```

the algorithm is going to produce 2 excessive rules:

```yaml
rules:
  - subset:
      kuma.io/service: frontend
      version: v1
    conf:
      action: Allow
  - subset:
      kuma.io/service: frontend
      version: !v1 # not 'v1'
    conf:
      action: Allow
```

The algorithm splits frontend service into `v1` and `not v1` but there is absolutely no reason for that. This happens because the current algorithm assumes that all provided rules could have an intersection. The new algorithm takes this into account. As a result, we're going to produce fewer rules, for example:

```yaml
type: MeshTrafficPermission
spec:
  targetRef:
    kind: Mesh
  from:
    - targetRef:
        kind: MeshService
        name: frontend_kuma-demo_svc_8080
      default:
        action: Allow
    - targetRef:
        kind: MeshServiceSubset
        name: backend_kuma-demo_svc_3001
        tags:
          team: mesh
          zone: east
          env: dev
      default:
        action: Deny
```

Old algorithm Inspect API:

image

New algorithm Inspect API:

image

### Checklist prior to review

- [X] [Link to relevant issue][1] as well as docs and UI issues --
- [X] This will not break child repos: it doesn't hardcode values (.e.g "kumahq" as a image registry) and it will work on Windows, system specific functions like `syscall.Mkfifo` have equivalent implementation on the other OS --
- [X] Tests (Unit test, E2E tests, manual test on universal and k8s) --
- [X] Do you need to update [`UPGRADE.md`](../blob/master/UPGRADE.md)? --
- [X] Does it need to be backported according to the [backporting policy](../blob/master/CONTRIBUTING.md#backporting)? --
- [X] Do you need to explicitly set a [`> Changelog:` entry here](../blob/master/CONTRIBUTING.md#submitting-a-patch) or add a `ci/` label to run fewer/more tests?

[1]: https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

### CODE REVIEW

**Review by lahabana — APPROVED** [review]

File: pkg/plugins/policies/core/rules/rules.go [hunk 1]
```diff
@@ -233,73 +279,107 @@ func BuildSingleItemRules(matchedPolicies []core_model.Resource) (SingleItemRule
...
 	}
 
 	sort.SliceStable(rules, func(i, j int) bool {
-		return rules[i].Subset.NumPositive() > rules[j].Subset.NumPositive()
+		ss1, ss2 := rules[i].Subset, rules[j].Subset
+		if ss1.NumPositive() != ss2.NumPositive() {
+			return ss1.NumPositive() > ss2.NumPositive()
+		}
+		h1, _ := json.Marshal(ss1)
+		h2, _ := json.Marshal(ss2)
```
Comment (jakubdyszkiewicz) [inline pkg/plugins/policies/core/rules/rules.go:381]: I'm slightly concerned about perf there. Isn't it called multiple times for every DPP on every XDS reconciliation?

File: pkg/plugins/policies/core/rules/rules_test.go [hunk 2]
```diff
@@ -141,6 +141,169 @@ var _ = Describe("Rules", func() {
...
+		type testCase struct {
+			s1, s2   core_rules.Subset
+			isSubset bool
+		}
+
+		DescribeTable("should respond if s2 is subset of s1",
+			func(given testCase) {
+				Expect(given.s1.IsSubset(given.s2)).To(Equal(given.isSubset))
+			},
+			Entry("", testCase{
```
Comment (jakubdyszkiewicz) [inline pkg/plugins/policies/core/rules/rules_test.go:154]: name the entries? at least "entry 1", "entry 2" etc. so we know which one fails

File: pkg/plugins/policies/core/rules/rules.go (hunk 1 above)
Comment (lahabana) [reply pkg/plugins/policies/core/rules/rules.go:381]: Yeah surely there's a better way to do a comparison here :)

File: pkg/plugins/policies/core/rules/rules.go [hunk 3]
```diff
@@ -73,10 +76,53 @@ func (ss Subset) IsSubset(other Subset) bool {
 			return false
 		}
 		for _, otherTag := range oTags {
-			if otherTag.Value == tag.Value && otherTag.Not != tag.Not {
+			if !singleDimSubset(tag, otherTag) {
```
Comment (michaelbeaumont) [inline pkg/plugins/policies/core/rules/rules.go:79]: nit but IMO would be easier to understand without the negation

File: pkg/plugins/policies/core/rules/rules.go [hunk 4]
```diff
@@ -233,73 +279,107 @@ func BuildSingleItemRules(matchedPolicies []core_model.Resource) (SingleItemRule
...
+			}
+			merged, err := MergeConfs(confs)
 			if err != nil {
 				return nil, err
 			}
-			if itemSubset.IsSubset(ss) {
-				confs = append(confs, item.GetDefault())
-				distinctOrigins[core_model.MetaToResourceKey(item.ResourceMeta)] = item.ResourceMeta
+			if merged != nil {
+				var origins []core_model.ResourceMeta
```
Comment (michaelbeaumont) [inline pkg/plugins/policies/core/rules/rules.go:359]: https://pkg.go.dev/golang.org/x/exp/slices#Clone

File: pkg/plugins/policies/core/rules/rules.go (hunk 3 above)
Comment (lobkovilya) [reply pkg/plugins/policies/core/rules/rules.go]: I renamed the method to `isSubset` so I think now it's understandable with negation

File: pkg/plugins/policies/core/rules/rules.go (hunk 4 above)
Comment (lobkovilya) [reply pkg/plugins/policies/core/rules/rules.go:371]: `distinctOrigins` is a map, so we can't clone it as slice

Comment (michaelbeaumont) [reply pkg/plugins/policies/core/rules/rules.go:371]: https://pkg.go.dev/golang.org/x/exp/maps#Values

**Review by jakubdyszkiewicz — APPROVED** [review]

**Review by michaelbeaumont — APPROVED** [review]

**Pull request merged**